## Let's read in some text and learn about tools available for working with strings
(these are some choice quotes from [Rear Admiral Grace Hopper](https://en.wikipedia.org/wiki/Grace_Hopper) -- inventor of the first compiler)

In [1]:
# First lets observe how we can loop through each line of the file
filename = 'gracehopperquotes.txt'
for line in open(filename, 'r'):
    print(line)


From then on, when anything went wrong with a computer, we said it had bugs in it.

If it's a good idea, go ahead and do it. It's much easier to apologize than it is to get permission.

It is often easier to ask for forgiveness than to ask for permission.

The most dangerous phrase in the language is, "We have always done it this way."

Humans are allergic to change. They love to say, "We have always done it this way." I try to fight that. That's why I have a clock on my wall that runs counter-clockwise.

A ship in port is safe, but that is not what ships are for. Sail out to sea and do new things.

You don't manage people, you manage things. You lead people.

Leadership is a two-way street, loyalty up and loyalty down. Respect for one's superiors; care for one's crew.

One accurate measurement is worth a thousand expert opinions.

I have always objected to doing anything over again if I had already done it once. 

Some day, on the corporate balance sheet, there will be an entry which 

## We can make a list out of the file using a loop too

In [2]:
text = []
for line in open(filename, 'r'):
    text.append(line) # this appends each line in the text to the end of the text list
print(text)

['From then on, when anything went wrong with a computer, we said it had bugs in it.\n', "If it's a good idea, go ahead and do it. It's much easier to apologize than it is to get permission.\n", 'It is often easier to ask for forgiveness than to ask for permission.\n', 'The most dangerous phrase in the language is, "We have always done it this way."\n', 'Humans are allergic to change. They love to say, "We have always done it this way." I try to fight that. That\'s why I have a clock on my wall that runs counter-clockwise.\n', 'A ship in port is safe, but that is not what ships are for. Sail out to sea and do new things.\n', "You don't manage people, you manage things. You lead people.\n", "Leadership is a two-way street, loyalty up and loyalty down. Respect for one's superiors; care for one's crew.\n", 'One accurate measurement is worth a thousand expert opinions.\n', 'I have always objected to doing anything over again if I had already done it once. \n', 'Some day, on the corporate b

## But now it has all those pesky '\n' characters in there! 
Those are new line characters that tell the computer a new line is starting. 

How can we get those out?

In [3]:
# loop through the items in the list
for i in range(len(text)):
    # i will loop through a list of numbers exactly as long as the number of items in text
    # so we can access each line by subscripting text with i
    # then we can use a string method replace('thing I want to get rid of;, 'thing I want to put in its place')
    # I could use .strip() here as well
    text[i] = text[i].replace('\n', '')
print(text)

['From then on, when anything went wrong with a computer, we said it had bugs in it.', "If it's a good idea, go ahead and do it. It's much easier to apologize than it is to get permission.", 'It is often easier to ask for forgiveness than to ask for permission.', 'The most dangerous phrase in the language is, "We have always done it this way."', 'Humans are allergic to change. They love to say, "We have always done it this way." I try to fight that. That\'s why I have a clock on my wall that runs counter-clockwise.', 'A ship in port is safe, but that is not what ships are for. Sail out to sea and do new things.', "You don't manage people, you manage things. You lead people.", "Leadership is a two-way street, loyalty up and loyalty down. Respect for one's superiors; care for one's crew.", 'One accurate measurement is worth a thousand expert opinions.', 'I have always objected to doing anything over again if I had already done it once. ', 'Some day, on the corporate balance sheet, there 

# More string methods

Some more methods that come in handy are upper(), and lower()
One reason you might want to use them is it helps find things 
no matter how they are capitalized



In [4]:
for i in range(len(text)):
    if 'bug' in text[i].lower():
        print(text[i])

From then on, when anything went wrong with a computer, we said it had bugs in it.


## Make some lists of your own
### 1) Make 1 list with all the quotes with the word "programming"

In [6]:
programming = []
for i in range(len(text)):
    if 'programming' in text[i].lower():
        programming.append(text[i])
print(programming)
# now use a loop to go through and test each line if the word you want is in it, then append it to your list

['To me programming is more than an important practical art. It is also a gigantic undertaking in the foundations of knowledge.']


 I only found one quote! What about you?
 
### 2) Now make a list with all the quotes that have 'humans' or 'people' or 'everyone' to get all of Grace's thoughts on people

In [13]:
people = []
for i in range(len(text)):
    if 'humans' in text[i].lower():
        people.append(text[i])
    elif 'people' in text[i].lower():
        people.append(text[i])
    elif 'everyone' in text[i].lower():
        people.append(text[i])
len(people)
# now use a loop to go through and test each line if the word you want is in it, then append it to your list

3

Now I got 3 quotes! (Hint: use len() to find out how many)

### Make a few more lists of your own!

### Now lets look at another file that has some numbers and text
#### This is a shopping list that is set up like this:
#### Item Number Price/each
* Item is a string
* Number is an integer
* Price/each is a float

In [14]:
filename = 'shoppinglist.txt'
for line in open(filename, 'r'):
    print(line)

Apples     100 2.0

Oranges    20  4.0

Kale       500 4.0

Oatmeal    25  1.0

Chicken    10  5.0

Honey      50  8.0

Tarantulas 2   3.0

Peanuts    300 0.5



## Now lets fill up three different lists with each of these pieces of information

In [15]:
# Heres some hints before we work on that
five = '5' # this is a string!
print(five + five) # addition doesnt work! because these are strings + just pushes these two strings together

55


In [16]:
# We can "cast" a type onto a variable and convert it into the type we want
reallyfive = int(five)
print(reallyfive + reallyfive)

10


In [18]:
# some more type casting
notafloat = '256.3'
reallyafloat = float(notafloat)
reallyafloat

256.3

In [19]:
# A useful method for strings is split(). If you give it no arguments it splits on the whitespace
alongstring = 'This string has several words and a number 7'
alist = alongstring.split()
print(alist)

['This', 'string', 'has', 'several', 'words', 'and', 'a', 'number', '7']


In [20]:
#Now we can subscript to get individual items, and then cast them to a different type!
seven = int(alist[-1])
print(seven*seven)

49


In [22]:
items = []
numtoget = []
price = []
for line in open(filename, 'r'):
    linelist= line.split()
    items.append(linelist[0])
    numtoget.append(linelist[1])
    price.append(linelist[2])
print(items, numtoget, price)
       # Your code goes here
        


['Apples', 'Oranges', 'Kale', 'Oatmeal', 'Chicken', 'Honey', 'Tarantulas', 'Peanuts'] ['100', '20', '500', '25', '10', '50', '2', '300'] ['2.0', '4.0', '4.0', '1.0', '5.0', '8.0', '3.0', '0.5']


## Make decisions and calculations

### So say I want to go buy everything EXCEPT the tarantulas and the peanuts. 
### How much is that all going to cost me? 
*(who put the tarantulas on the list anway?)*

In [25]:
total = 0
for i in range(len(items)):
    thing = items[i].lower()
    if thing != 'tarantulas':
        if thing != 'peanuts':
            total= total + int(numtoget[i])*float(price[i])
total
            #calculate the cost of getting the number of things at this index times the price/each

2755.0

I got 2,755$ bucks!

## Geez, that's way too much. What if I only want at most 25 of any single item?
(still no peanuts or tarantuals please)

*(So go through and test for the tarantuals and peanuts again, then test the numtoget[i] and if it is over 25, only get 25, otherwise just get the number asked for.)*

In [29]:
total = 0
for i in range(len(items)):
    thing = items[i].lower()
    if thing != 'tarantulas':
        if thing != 'peanuts':
            if int(numtoget[i]) <= 25:
                total= total + int(numtoget[i])*float(price[i])
            else:
                total= total+ 25*float(price[i])
total
# your code goes here

505.0

505$ That's still alot, but it's more reasonable. I guess I'll go with that.



In [30]:
# All done!
import comp116
comp116.submit('ws9')